### PR TITLE
Add go client

### PR DESCRIPTION
**Pull request by schrodit**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
cc @dguendisch @Diaphteiros 

**Special notes for your reviewer**:

**Release note**:

```improvement operator

```

### CODE REVIEW

File: client-go/apis/v2/overwrite.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+type OverwriteDeclaration struct {
+	// DeclaringComponent uniquely identifies the component
+	DeclaringComponent ComponentMetadata `json:"declaringComponent"`
+
+	// Overwrites defines the resolvable component overwrites declared
+	// by a component
+	Overwrites []ComponentOverwrite `json:"overwrites"`
+}
+
+// ComponentOverwrite defines a overwrite for a resolvable component.
```
Comment (dguendisch) [inline client-go/apis/v2/overwrite.go:29]:
```suggestion
// ComponentOverwrite defines an overwrite for a resolvable component.
```

File: client-go/apis/v2/overwrite.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	// ComponentOverwrites will overwrite overwritable attributes from the
+	// component itself.
+	ComponentOverwrites json.RawMessage `json:"componentOverwrites"`
+
+	// DependencyOverwrites defines the overwrites for the components dependencies.
+	DependencyOverwrites ComponentList `json:"dependencyOverwrites,omitempty"`
+}
+
+// ApplyOverwrites applies the overwrites defined by component descriptor
+// and returns a new component descriptor with the applies overwrites
```
Comment (dguendisch) [inline client-go/apis/v2/overwrite.go:43]:
```suggestion
// and returns a new component descriptor with the applied overwrites
```

File: client-go/apis/v2/componentdescriptor.go [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+type ComponentAccessor interface {
+	// GetMetadata returns the metadata of the component
+	GetMetadata() ComponentMetadata
+	// SetMetadata sets the component metadata.
+	SetMetadata(metadata ComponentMetadata)
+	// GetAdditionalData returns the custom data of a component.
+	GetAdditionalData() ([]byte, error)
+	// SetAdditionalData sets the custom data of a component.
+	SetAdditionalData([]byte) error
+	// ApplyOverwrite should apply a component overwrite and return the new overwritten component.
```
Comment (dguendisch) [inline client-go/apis/v2/componentdescriptor.go:69]:
```suggestion
	// ApplyOverwrite applies a component overwrite and returns the new overwritten component.
```

File: client-go/apis/v2/resolvable.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+// ResolvableComponent describes a top-level component with its dependencies
+type ResolvableComponent struct {
+	ComponentMetadata `json:",inline"`
+
+	// Dependencies contain all components that the component depends on.
+	Dependencies ComponentList `json:"dependencies,omitempty"`
+}
+
+// ApplyOverwriteToDependencies applies the given overwrite of the dependencies of a resolvable component
+// and return the new resolvable component.
```
Comment (dguendisch) [inline client-go/apis/v2/resolvable.go:30]:
```suggestion
// and returns the new resolvable component.
```

File: client-go/apis/v2/types.go [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+package v2
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/ghodss/yaml"
+)
+
+// OCIImageType is the type of a oci image component
```
Comment (dguendisch) [inline client-go/apis/v2/types.go:24]:
```suggestion
// OCIImageType is the type of an oci image component
```

File: client-go/apis/v2/types.go [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+}
+
+func (O OCIImage) DeepCopy() ComponentAccessor {
+	img := &OCIImage{}
+	img.SetMetadata(O.GetMetadata())
+	img.ImageReference = O.ImageReference
+	return img
+}
+
+// OCIImageType is the type of a oci image component
```
Comment (dguendisch) [inline client-go/apis/v2/types.go:93]:
```suggestion
// OCIImageType is the type of an oci image component
```

File: client-go/apis/v2/types.go [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+	//if err := yaml.Unmarshal(c.GetAdditionalData(), &data); err != nil {
+	//	return nil, err
+	//}
+	//
+	//// replace addData in data
+	//
+	//newAddData, err := yaml.Marshal(data)
+	//if err != nil {
+	//	return nil, err
+	//}
```
Comment (dguendisch) [inline client-go/apis/v2/types.go:211]: remove?

File: client-go/apis/v2/types_resolvable.go [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+}
+
+func (g GardenerComponent) DeepCopy() ResolvableComponentAccessor {
+	comp := &GardenerComponent{}
+	comp.SetMetadata(g.GetMetadata())
+	comp.SetDependencies(g.GetDependencies())
+	return comp
+}
+
+// GardenerComponentType is the type of a gardener component
```
Comment (dguendisch) [inline client-go/apis/v2/types_resolvable.go:89]: Copy pasta :)

File: client-go/apis/v2/types_resolvable.go [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+	ResolvableSchemeEncoder: ResolvableEncoderFunc(func(accessor ResolvableComponentAccessor) ([]byte, error) {
+		comp, ok := accessor.(*OCIComponent)
+		if !ok {
+			return nil, fmt.Errorf("accessor is not of type %s", OCIImageType)
+		}
+		return yaml.Marshal(comp)
+	}),
+}
+
+// OCIComponent describes oci resolvable component
```
Comment (dguendisch) [inline client-go/apis/v2/types_resolvable.go:109]:
```suggestion
// OCIComponent describes an oci resolvable component
```

File: client-go/apis/v2/validation/validation_test.go [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+			},
+			ImageReference: "docker/image1:1.2.3",
+		}
+		ociImage2 = &v2.OCIImage{
+			ComponentMetadata: v2.ComponentMetadata{
+				Type:    v2.OCIImageType,
+				Name:    "image2",
+				Version: "1.2.3",
+			},
+			ImageReference: "docker/image1:1.2.3",
```
Comment (dguendisch) [inline client-go/apis/v2/validation/validation_test.go:55]: copy pasta? you meant `image2`?

**Comment by schrodit** [comment]
closed in favor of https://github.com/gardener/component-spec/pull/2

**Pull request closed without merging**